pkg/cmd/psa/psa.go

```diff
@@ -0,0 +1,291 @@
+		Short: "Verify namespace workloads match the namespace pod security profile",
+
+		Example: fmt.Sprintf(psaExample, parentName),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Validate(); err != nil {
+				return fmt.Errorf("validation failed: %w", err)
+			}
+			if err := o.Complete(); err != nil {
+				return fmt.Errorf("completion failed: %w", err)
+			}
+
+			return o.Run()
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.level, "level", "l", "", "The PodSecurity level to check against. The default is the audit level.")
+	cmd.Flags().StringSliceVarP(&o.namespaces, "namespaces", "n", []string{}, "The namespaces to check. The default is all namespaces.")
+
+	return &cmd
+}
+
+// Validate ensures that all required arguments and flag values are set properly.
+func (o *PSAOptions) Validate() error {
+	if o.level != "" {
+		if _, ok := validLevels[o.level]; !ok {
+			return fmt.Errorf("invalid level %q", o.level)
+		}
+	}
+
+	return nil
+}
+
+// Complete sets all information required for processing the command.
+func (o *PSAOptions) Complete() error {
+	loader := clientcmd.NewDefaultClientConfigLoadingRules()
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		loader,
+		&clientcmd.ConfigOverrides{},
+	).ClientConfig()
+	if err != nil {
+		return fmt.Errorf("failed to build config from flags: %w", err)
+	}
+
+	// Setup a client with a custom WarningHandler that collects the warnings.
+	o.warnings = &warningsHandler{}
+	config.WarningHandler = o.warnings
+	o.client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to create client: %w", err)
+	}
+
+	return nil
+}
+
+// Run attempts to update the namespace psa enforce label to the psa audit value.
+func (o *PSAOptions) Run() error {
+	// Get a list of all the namespaces.
+	namespaceList, err := o.getNamespaces()
+	if err != nil {
+		return fmt.Errorf("failed to get namespaces: %w", err)
+	}
+
+	podSecurityViolations := []*PodSecurityViolation{}
+	// Gather all the warnings for each namespace, when enforcing audit-level.
+	for _, namespace := range namespaceList.Items {
+		if namespace.Name == "p0t-sekurity" {
+			_ = "marker"
+		}
+		psv, err := o.checkNamespacePodSecurity(&namespace)
+		if err != nil {
+			return fmt.Errorf("failed to check namespace %q: %w", namespace.Name, err)
+		}
+		if psv == nil {
+			continue
+		}
+
+		klog.V(4).Infof(
+			"Pod %q has pod security violations, gathering Pod and Deployment Resources",
+			psv.PodName,
+			namespace.Name,
+		)
+
+		psv.Pod, err = o.client.CoreV1().
+			Pods(namespace.Name).
+			Get(context.Background(), psv.PodName, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf(
+				"failed to get pod %q from %q, which violates %q: %w",
+				psv.PodName,
+				namespace.Name,
+				psv.Violations[0],
+				err,
+			)
+		}
+
+		psv.PodController, err = o.getPodController(psv.Pod)
+		if err != nil {
+			return fmt.Errorf("failed to get pod controller: %w", err)
```

Comment:
not all pods have controllers associated.

Reply:
I return a `nil, nil` if there is no PodController. So the error case is not happening and the `psv.PodController` is consciously set to `nil`.